### ISSUE TITLE
get_llvm_tool_names relies on hardcoded, outdated suffixes list

### ISSUE THREAD

**Issue by amyspark** [mesonbuild/meson#12145]
**Describe the bug**

Meson's `get_llvm_tool_names` relies on a hardcoded, outdated list of versions to create the suffixes. This is easily broken, as LLVM has already released v17 and v18, and existing releases aren't forward compatible, thus breaking users of those compilers.

https://github.com/mesonbuild/meson/blob/253df6f9f821a3052b56673d5427aca9dfe41eba/mesonbuild/environment.py#L152-L160

**To Reproduce**

Run `meson coverage` against a build that has been configured with `CC=clang-18` `CXX=clang-18`, eg. https://github.com/blackmagic-debug/bmpflash/tree/feature/continuous-integration.

**Expected behavior**

Meson should work out of the box, nomatter the release of Clang that's being used.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? No
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) GHA Ubuntu 22.04
* what Python version are you using e.g. 3.8.0  3.10.12
* what `meson --version` 1.2.1
* what `ninja --version` if it's a Ninja build 1.11.1.git.kitware.jobserver-1

**Comment by tristan957** [comment]
It seems like this could be done more heuristically indeed, but how would one know the upper bound? Thanks for bringing this up.

**Comment by dragonmux** [comment]
The heuristic we'd apply is "is the suffix numeric" - any attempt at an upper bound is likely to fail sometime off into the future, but LLVM/Clang have been very good to maintain suffixing with the numeric major version of the release like clockwork, and that's probably dependable into the future.

**Comment by tristan957** [comment]
Would one of y'all be interested in submitting a PR to make this more flexible?

**Comment by eli-schwartz** [comment]
> The heuristic we'd apply is "is the suffix numeric" - any attempt at an upper bound is likely to fail sometime off into the future, but LLVM/Clang have been very good to maintain suffixing with the numeric major version of the release like clockwork, and that's probably dependable into the future.

I'm not sure how this even constitutes a heuristic to recommend. Your objection is that meson guesses at a hardcoded list of names to attempt to find existing llvm-config tools, and your recommendation is that given the path to an llvm-config tool we should simply verify that the suffix is numeric?

But the entire problem is we do not *have* the path to an llvm-config tool. If we can find an unversioned one where we don't have to guess at trials to try, we prefer that.

But if we can't find the unversioned one, what are we supposed to do? Start with llvm-config-9999999999 and then enumerate every number in the Hindu-Arabic numeral system, each time stat'ing the filesystem along the length of $PATH to see whether such a binary exists?

**Comment by eli-schwartz** [comment]
If you have a better heuristic than the one suggested so far in this ticket, I would be very thrilled to make meson better about this. But please, before "submitting a PR to make this more flexible", it would be great to have a discussion / explanation about what would be a good approach to take.

**Comment by amyspark** [comment]
> it would be great to have a discussion / explanation about what would be a good approach to take

Sorry, I have to disagree here. With a PR, one could at least compare and contrast the proposed approach and its shortcomings. Waiting for such a design discussion to conclude would instead cause this issue to become as stagnant as e.g. #11123, #11124, #11641 (those of my own, I'm sure there are plenty more in the issue and MR trackers).

**Comment by eli-schwartz** [comment]
I don't understand the comparisons you're trying to make. 

Furthermore, I'm not forbidding you from writing code. (I am not sure how I could be forbidding that -- meson doesn't require opening an issue before submitting a PR, even.) I am *recommending* as a friendly suggestion that you put into words, exactly what you want here -- something you have to do by the very nature of trying to write that code anyway.

Since I apparently failed to put my concern into words before, let me try again:

I'm not sure there's a bug here. Could you explain why you think this is a bug, rather than an unavoidable need imposed by upstream naming conventions? I'm also worried that trying to solve a problem without even understanding what the problem *is*, will end up being fruitless.

**Comment by nirbheek** [comment]
> Start with llvm-config-9999999999 and then enumerate every number in the Hindu-Arabic numeral system, each time stat'ing the filesystem along the length of $PATH to see whether such a binary exists?

No, you'd do the reverse. :slightly_smiling_face: Scan `PATH` for binaries matching `llvm-config-*` and pick the largest numerical suffix.

**Comment by nirbheek** [comment]
> If the user has not specified what compiler to use and we pick clang, or if they have, it's suffixed, and we can't find the suffixed LLVM tools, then Meson _must_ get the version information out of that clang and find a matching LLVM

This also makes sense to me, but llvm-config is also used when we're not using clang, so we probably need to handle that case too.

**Comment by eli-schwartz** [comment]
> No, you'd do the reverse. :slightly_smiling_face: Scan `PATH` for binaries matching `llvm-config-*` and pick the largest numerical suffix.

That sounds quite a bit like implementing lots of `shutil.which`, including handling path extensions (and only path extensions!) as Windows-specific modifiers to "a numerical suffix". Then repeating all that work when running `shutil.which`. Or outright implementing a drop-in replacement for that function, which remembers and sorts which candidates it found per directory and then re-iterating that mapping based on number first, directory second.

**Comment by eli-schwartz** [comment]
> the crux of what's going on here is that given `CC=clang-18 CXX=clang++-18`, Meson *ignores* the suffix, guessing at random LLVMs to try till it picks one (very probably not the correct one, though if it's the only Clang+LLVM and there are unsuffixed versions it'll by chance get things right).

That's fair, albeit clang-specific. Guessing at random LLVMs is the only way to do it for GCC.

> A failure to find matching Clang+LLVM should result in a failure to meson setup.

I'm not sure how strict we should be here. At least if the user manually specifies a given `$LLVM_CONFIG` that is older than `$CXX` they may have legitimate reasons to do so -- i.e. building with the clang compiler at latest because a newer compiler is simply better, but requiring a specific version of LLVM due to the notoriously unstable API that frequently requires rewriting your codebase to support a specific LLVM version.

When hitting the "search for a suitable llvm-config through guesswork" path I can hear arguments both for "first guess based on the clang version" and for "require a match against the clang version". I think either one is probably a reasonable interpretation, so I guess you should pick whichever one you feel is better.

**Comment by eli-schwartz** [comment]
Also -- thanks for the additional analysis and description of what the exact problem is that we're trying to solve. ;)

**Comment by eli-schwartz** [comment]
> The current approach Meson uses is to [pot-shot guess at LLVMs to use](https://github.com/mesonbuild/meson/blob/da78b969ee3635bbea81909aa9ebdad3c77257a5/mesonbuild/environment.py#L152-L184) without any kind of checking about suitability. This results in extremely broken behaviour and is the reason we had to uninstall all older LLVMs from our CI environment for [bmpflash](https://github.com/blackmagic-debug/bmpflash) - like [in this build](https://github.com/blackmagic-debug/bmpflash/actions/runs/5931031441/job/16082033330#step:11:24).

This is an interesting error:
> ```
> llvm-ar: error: deps/substrate/impl/libsubstrate.a: Opaque pointers are only supported in -opaque-pointers mode (Producer: 'LLVM17.0.0' Reader: 'LLVM 14.0.0')
> ```

I think this means that llvm-ar from PATH is from a system llvm 14, while you compiled with clang 17... and the problematic issue isn't llvm-config at all, it is `$AR`. But clang is, as I recall, able to handle old producers (clang version) with a new reader (ar, lld), and the problem is when you do the reverse. So if your build VM has a default clang that's really old, and you install a newer one and invoke it directly with `CC=clang-18 CXX=clang++-18`, you can end up with a too-old llvm-ar.

This is handled here: https://github.com/mesonbuild/meson/blob/bde690b06e930020a0ec6ccaea7a76babf77dff5/mesonbuild/compilers/detect.py#L179-L184

At the very least,
```python
[defaults['clang_static_linker']]
```
should probably be 
```python
[defaults['clang_static_linker'] + '-' + compiler.version.split('.')[0], defaults['clang_static_linker']]
```

**Comment by dragonmux** [comment]
> I'm not sure how strict we should be here. At least if the user manually specifies a given $LLVM_CONFIG that is older than $CXX they may have legitimate reasons to do so -- i.e. building with the clang compiler at latest because a newer compiler is simply better, but requiring a specific version of LLVM due to the notoriously unstable API that frequently requires rewriting your codebase to support a specific LLVM version.

This cannot work. The only way for any of that to work is an older Clang and a newer LLVM. Even that's sketchy as heck. This is beautifully demo'd by the build failure we referenced earlier.

> I think this means that llvm-ar from PATH is from a system llvm 14, while you compiled with clang 17... and the problematic issue isn't llvm-config at all, it is $AR.

Partially incorrect (especially as we do not set $AR to anything). [There are problems](https://github.com/blackmagic-debug/bmpflash/actions/runs/5931349235/job/16083042452#step:11:22) caused by Meson feeding from the wrong llvm-config a bad `LLVMgold-N.so` to the binutils ld - this is an LTO'd build and has problems with the LTO plugin, as well as llvm-ar. It is endemic to the problem we're talking about here though - Meson fails, point blank, to keep the proper lockstep between Clang and LLVM leading to these strange problems caused by it creating broken environments.

Do note that none of the problems outlined here happened for the Clang 13 build specifically because that's the right way around for Clang vs LLVM newness; and that after removing the other LLVMs, none of the badness happened any more for anything other than Clang 17 which Meson doesn't know about yet.

That we can tell, GHA doesn't ship defaults packages for LLVM and Clang as you are supposed to pick which you want to use by environment variable - as we have done. We can very well prove that Meson got it wrong here because of the environment.py code that guesses randomly as it doesn't understand what the -17 suffix is yet.

**Comment by eli-schwartz** [comment]
> This cannot work. The only way for any of that to work is an older Clang and a newer LLVM. Even that's sketchy as heck. This is beautifully demo'd by the build failure we referenced earlier.

I'm pretty sure this is not correct... The problem is that if you are going to LTO two different versions of LLVM IR together (including e.g. some IR that is contained inside of libLLVM?), you need to use the newer toolchain. Your version of libLLVM cannot be newer than your toolchain.

Although maybe the source of the confusion here is about LLVM as a toolchain vs. LLVM as a dependency library.

- dependency('llvm') / llvm-config programs
- llvm-binutils software package provides toolchain binaries e.g. llvm-ar, LLVMgold.so etc
- llvm-clang software package provides toolchain binaries e.g. clang driver
- llvm-lld software package provides toolchain binaries e.g. lld


Note that Meson only uses llvm-config for the first of these -- for dependency().

> Partially incorrect (especially as we do not set $AR to anything). [There are problems](https://github.com/blackmagic-debug/bmpflash/actions/runs/5931349235/job/16083042452#step:11:22) caused by Meson feeding from the wrong llvm-config a bad `LLVMgold-N.so` to the binutils ld - this is an LTO'd build and has problems with the LTO plugin, as well as llvm-ar. It is endemic to the problem we're talking about here though - Meson fails, point blank, to keep the proper lockstep between Clang and LLVM leading to these strange problems caused by it creating broken environments.

I am not sure what the answer to this is, if you have a really old LLVMgold.so plugin and want to use a custom installed toolchain which is much newer and still use the BFD linker. You could try setting -fuse-ld=lld and clang should pick up the lld shipped internally which has matching LTO compat.

And I'm not partially incorrect about the AR thing -- I'm saying the error you get here is because meson is dumb and assumes that if you manually set CXX to a *versioned* compiler, you must also set $AR to a *versioned* archiver to match.

... at least if you want LTO to work.

> We can very well prove that Meson got it wrong here because of the environment.py code that guesses randomly as it doesn't understand what the -17 suffix is yet.

Impossible, since the environment.py code that guesses randomly isn't used by llvm-ar detection, it is only used for llvm-config detection (and for coverage tools, clang-format, scan-build etc.)

You can change environment.py and add the -17 suffix but it simply will not fix an "llvm-ar" error. It should fix the `ninja coverage` issue by correctly detecting llvm-cov.

**Comment by dragonmux** [comment]
> Although maybe the source of the confusion here is about LLVM as a toolchain vs. LLVM as a dependency library.

No, the source of confusion isn't in mixing libLLVM and LLVM, it's in not knowing the innards in Meson like you do and being told "oh, that's llvm-config in play" or w/e and then later being told "oh, but not here". We know exactly how the `llvm-N` and `llvm-linker-tools-N` packages are laid out and interact here with the compilation process. We're working out how Meson tries to make them play, and how it gets that wrong.

> Impossible, since the environment.py code that guesses randomly isn't used by llvm-ar detection, it is only used for llvm-config detection (and for coverage tools, clang-format, scan-build etc.)

Please, pray tell then, where on earth is *Meson* getting the wrong `LLVMgold-N.so` from then? Because the one it's picking to tell ld.bfd about is completely wrong.

> if you have a really old LLVMgold.so plugin and want to use a custom installed toolchain which is much newer and still use the BFD linker

The CI environment installs LLVMgold-17.so and then Meson tells ld.bfd to use the old one through some mechanism we haven't yet figured out. That's how this linker error occurs. We've not told it to do any of that.

----

We could have a back and forth all day, but we've instead been exploring how compiler and linker detection is done and started writing some patches. Addressing `llvm-ar` was actually remarkably easy - https://github.com/dragonmux/meson/commit/3e2d0fa7e1a7e819e429821bf7744af9f27df933

The coverage backend however is wilfully ignoring the coredata and doesn't get passed any of that, so fixing that'll take a bit more. Same for clang-tidy and clang-format. Given `llvm-cov` version is tied strongly to the `clang` version, that's more problematic. The others will survive by comparison.

Despite being post-setup phase, is there an objection to us pulling open the coredata so we can patch the coverage stuff right rather than adding to the mound of hacks? This is specifically so we can access CoreData.compilers to pull out the proper information to locate a proper llvm-cov with.

**Comment by eli-schwartz** [comment]
> Please, pray tell then, where on earth is _Meson_ getting the wrong `LLVMgold-N.so` from then? Because the one it's picking to tell ld.bfd about is completely wrong.

> The CI environment installs LLVMgold-17.so and then Meson tells ld.bfd to use the old one through some mechanism we haven't yet figured out. That's how this linker error occurs. We've not told it to do any of that.

I am not sure why you think meson is getting any LLVMgold-N.so at all -- meson simply passes `-flto` and the clang linker driver internally handles the plugin.

But this is very easy to verify. Stop worrying about the innards of meson. Just run ninja with the `-v` option to print the used command line, to check how clang is *actually* being invoked.

Once you know the command line in use, you can even try running that command line a few times manually in a shell script with additional debugging information such as `clang++-17 -v`.



> We could have a back and forth all day, but we've instead been exploring how compiler and linker detection is done and started writing some patches. Addressing `llvm-ar` was actually remarkably easy - [dragonmux@3e2d0fa](https://github.com/dragonmux/meson/commit/3e2d0fa7e1a7e819e429821bf7744af9f27df933)

Awesome, thanks!



> Despite being post-setup phase, is there an objection to us pulling open the coredata so we can patch the coverage stuff right rather than adding to the mound of hacks? This is specifically so we can access CoreData.compilers to pull out the proper information to locate a proper llvm-cov with.

The usual way we do this with other ninja rules implemented inside of `meson --internal` scripts is by adding a command-line option to the script and passing the path to e.g. llvm-cov as detected at configure time. See for example the gettext, msgfmt, gtkdochelper scripts inside of mesonbuild/scripts

I would prefer this instead of loading the coredata, in part because the coredata tends to drag in basically the entire meson codebase as a runtime dependency of the code opening it, whereas these tiny internal scripts are, preferably, fast to start up and have minimal dependencies.

**Comment by dragonmux** [comment]
> The usual way we do this with other ninja rules implemented inside of meson --internal scripts is by adding a command-line option to the script and passing the path to e.g. llvm-cov as detected at setup time. See for example the gettext, msgfmt, gtkdochelper scripts inside of mesonbuild/scripts

We'll look at this alternative way then - disentangling how the script tries to re-discover the coverage tools is going to be pain, but that's a fair point. That means that not only do we need to replace how it finds `llvm-cov`, but all of the coverage tools. Right now, that script calls `environment.find_coverage_tools()` at the start of its main function's execution, and therein lies the problem we were aiming to solve with the coredata.

**Comment by dragonmux** [comment]
We have done our best to implement the suggested paradigm, please give the linked PR A look, Eli.
